codes/python/chapter_tree/avl_tree.py

```diff
@@ -0,0 +1,281 @@
+            val: 要插入的结点的值
+
+        Returns: 插入结点后的根结点
+
+        """
+        if node is None:
+            return AVLTreeNode(val)
+        # 1. 查找插入位置，并插入结点
+        if val < node.val:
+            node.left = self.insert_helper(node.left, val)
+        elif val > node.val:
+            node.right = self.insert_helper(node.right, val)
+        else:
+            # 重复结点不插入，直接返回
+            return node
+        # 更新结点高度
+        self.__update_height(node)
+        # 2. 执行旋转操作，使该子树重新恢复平衡
+        return self.rotate(node)
+
+    def remove(self, val: int):
+        """
+        删除结点
+        Args:
+            val: 要删除的结点的值
+
+        Returns:
+
+        """
+        root = self.remove_helper(self.root, val)
+        return root
+
+    def remove_helper(
+        self, node: typing.Optional[AVLTreeNode], val: int
+    ) -> typing.Optional[AVLTreeNode]:
+        """
+        递归删除结点（辅助函数）
+        Args:
+            node:  删除的起始结点
+            val: 要删除的结点的值
+
+        Returns: 删除目标结点后的起始结点
+
+        """
+        if node is None:
+            return None
+        # 1. 查找结点，并删除之
+        if val < node.val:
+            node.left = self.remove_helper(node.left, val)
+        elif val > node.val:
+            node.right = self.remove_helper(node.right, val)
+        else:
+            if node.left is None or node.right is None:
+                child = node.left or node.right
+                # 子结点数量 = 0 ，直接删除 node 并返回
+                if child is None:
+                    return None
+                # 子结点数量 = 1 ，直接删除 node
+                else:
+                    node = child
+            else:  # 子结点数量 = 2 ，则将中序遍历的下个结点删除，并用该结点替换当前结点
+                temp = self.min_node(node.right)
+                node.right = self.remove_helper(node.right, temp.val)
+                node.val = temp.val
+        # 更新结点高度
+        self.__update_height(node)
+        # 2. 执行旋转操作，使该子树重新恢复平衡
+        return self.rotate(node)
+
+    def min_node(
+        self, node: typing.Optional[AVLTreeNode]
+    ) -> typing.Optional[AVLTreeNode]:
+        # 获取最小结点
+        if node is None:
+            return None
+        # 循环访问左子结点，直到叶结点时为最小结点，跳出
+        while node.left is not None:
+            node = node.left
+        return node
+
+    def search(self, val: int):
+        cur = self.root
+        while cur is not None:
+            if cur.val < val:
+                cur = cur.right
+            elif cur.val > val:
+                cur = cur.left
+            else:
+                break
+        return cur
+
+
+if __name__ == "__main__":
+
+    def test_insert(tree: AVLTree, val: int):
+        tree.insert(val)
+        print("\n插入结点 {} 后，AVL 树为".format(val))
+        print_tree(tree.root)
+
+    def test_remove(tree: AVLTree, val: int):
+        tree.remove(val)
+        print("\n删除结点 {} 后，AVL 树为".format(val))
+        print_tree(tree.root)
+
+    # 初始化空 AVL 树
+    avl_tree = AVLTree()
+
+    # 插入结点
+    # 请关注插入结点后，AVL 树是如何保持平衡的
+    test_insert(avl_tree, 1)
```

Comment:
`print_tree(avl_tree)` is required after each operation. Please check the correspondence to Java code carefully.